src/proxy.c

```diff
@@ -238,113 +239,131 @@ cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
 	msg.msg_control = ctl_buffer;
 	msg.msg_controllen = sizeof(ctl_buffer);
 
-	/* read length */
-	bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
-
-	if (bytes_read == -1){
-		g_critical("failed to read proxy message %s", strerror(errno));
-		return false;
+	/* 
+	 * This sleep is very important to make sure we don't
+	 * get MSG_CTRUNC message from msghdr. There is no way
+	 * to recover from a MSG_CTRUNC the control message lost.
+	 */
+	if (proxy_data->oob_fd) {
+		usleep(1e3);
 	}
 
-	if (bytes_read == 0) {
-		g_critical("failed to read lenght");
-		return false;
-	}
+	/* read at least message length */
+	for (timeout = 0; timeout < 100; timeout++) {
+		if (tot_b_rd >= sizeof(buf_len)) {
+			break;
+		}
 
-	if (proxy_data->oob_fd) {
-		/* check if oob data was received */
-		cmsg = CMSG_FIRSTHDR(&msg);
-		if (cmsg) {
-			data = CMSG_DATA(cmsg);
-			if (data) {
-				*proxy_data->oob_fd = *((int*) data);
-				g_message("oob fd read from proxy socket %d",
-					*proxy_data->oob_fd);
-				fd_read = true;
-			}
+		bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
+		if (bytes_read < 1) {
+			usleep(1e3);
+			continue;
 		}
-	}
 
-	/* only out-of-band data was received */
-	if (bytes_read < sizeof (buf_len) || iov_buffer[0] == 'F' ) {
-		goto out1;
+		tmp_ptr = realloc(frame, tot_b_rd + (size_t)bytes_read);
+		if (!tmp_ptr) {
+			g_critical("failed to realloc memory");
+			goto out;
+		}
+		frame = tmp_ptr;
+		memcpy(frame + tot_b_rd, iov_buffer, (size_t)bytes_read);
+		tot_b_rd += (size_t)bytes_read;
 	}
 
-	/* cc-proxy sends a fd as out-of-band data
-	 * followed by an 'F'.
-	 * *DO NOT* copy these bytes to msg_received
-	 */
-	memcpy (buf_len, iov_buffer, sizeof (buf_len));
-	p_msg.length = cc_oci_get_big_endian_32 (buf_len);
-	g_debug ("msg length: %d", p_msg.length);
+	if (tot_b_rd < sizeof(buf_len)) {
+		g_critical("failed to get message length, only %ld bytes read", tot_b_rd);
+		goto out;
+	}
 
-	/* ensure DO NOT copy extra bytes */
-	bytes_to_copy = MIN ((size_t)bytes_read-data_offset, (size_t)p_msg.length);
-	g_string_append_len (proxy_data->msg_received,
-		iov_buffer+data_offset, (ssize_t)bytes_to_copy);
+	/* compute message length */
+	memcpy (buf_len, frame, sizeof(buf_len));
+	p_msg.length = cc_oci_get_big_endian_32(buf_len);
+	g_debug("msg length: %d", p_msg.length);
 
-	/* all message was read */
-	if (bytes_read >= p_msg.length+data_offset) {
-		goto out2;
+	tot_b_expected = p_msg.length + data_offset;
+	if (proxy_data->oob_fd) {
+		tot_b_expected += 1;
 	}
 
-	/* read missing bytes */
-	while(true) {
-		bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
-		if (bytes_read < 0) {
-			/* no more data to read */
+	/* read remaining bytes */
+	for (timeout = 0; timeout < 100; timeout++) {
+		if (tot_b_rd >= tot_b_expected) {
 			break;
 		}
-		if (! proxy_data->oob_fd) {
-			/* this message has not out-of-band data */
+
+		bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
+		if (bytes_read < 1) {
+			usleep(1e3);
```

Comment:
No, I am not waiting for OOB data specifically, just a sleep to avoid CPU overload.